### PR TITLE
Auto populate primary program plan on program enrollment

### LINKED ISSUE #401: Ability to link multiple courses to multiple programs and multiple programs to multiple courses (include Prerequisite)
Power of US Idea Owner: [Adeline Moore](https://powerofus.force.com/_ui/core/userprofile/UserProfilePage?u=0058000000A53LhAAJ&tab=sfdc.ProfilePlatformOverview)
[Power of US Idea](https://powerofus.force.com/hub-ideas/1140)

**Enhancement Request**: We would love to see the courses linked to Academic Programs as prerequisites for the Program. Also we need to link multiple courses to programs. Right now I do not see a way of doing this.

**Requirements:** 

- Ability to associate 1 or more Courses to 1 or more Academic Programs
- Ability to associate 1 or more Academic Programs to 1 or more Courses
- All Courses linked to an Academic Program are marked as a prerequisites for the Academic Program
- Reporting should include the ability to:
- Search on an Course and see all associated Academic Programs
- Search on an Academic Program and see all Associated Courses

**Update & Summary:**
This is the ability to link multiple courses to multiple programs or vice versa to achieve the degree associated with an Academic Program (Account).  If a particular course (course offering) is required for two separate Academic Programs (e.g. BIO 101 for both BS in Biology and BS in Biomedical Engineering) it can be linked to both. On the other end, if you get to a point in the Academic Program where you have to take a language course (for example), you have multiple courses to draw from that fulfill that plan requirement. 

This work will result in 2 new objects, "Program Plan" and "Plan Requirement", that will sit between Academic Program (Account), Program Enrollment, and Course. There will also be a new field on the existing Program Enrollment" to reference "Plan" (aka list or sequence of courses a student is following).

- **Originally from a PUB Idea**: https://powerofus.force.com/hub-ideas/1140
- **Blog Post on the topic**: https://salesforcelady.wordpress.com/2017/01/04/how-to-link-courses-and-academic-programs-in-heda/#more-1146

### LINKED ISSUE #662: Default Program Plan when creating Program Enrolment
Hi there,

I noticed that the program plan is no longer be populated automatically when a program enrolment is created. Has this been removed?

### PR DESCRIPTION
**Pull request by xinnanlisf**
# Critical Changes

# Changes
- We've introduced a new TDTM class, PREN_ProgramPlan_TDTM, that automatically populates the Program Plan field on a Program Enrollment. Now, when you create a Program Enrollment and leave the Program Plan field blank, we set it to the one that's specified as Primary for the Academic Program. If you haven't specified a Primary Program Plan, we don't fill in a value.

# Issues Closed
Fixes #662
# New Metadata

# Deleted Metadata

# Testing Notes
1. Create a academic program
2. Create a program plan, mark it as primary and assign it to the academic program
3. Create a contact
4. Create a program enrollment and set the account as the academic program
5. Make sure the program plan on program enrollment is populated with the primary program plan

### CODE REVIEW

File: src/classes/PREN_ProgramPlan_TDTM.cls [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+                                    WHERE Account__c in :setAcademicProgramIds
+                                        AND Is_Primary__c = TRUE]) {
+            if (pp.Account__c != null) {
+                mapAcademicProgramIdToPrimaryPPId.put(pp.Account__c, pp.Id);
+            }
+        }
+
+        //Loop through program enrollment again and populate primary program plan
+        for (SObject so : newlist) {
+            Program_Enrollment__c newPe = (Program_Enrollment__c) so;
```
Comment (lifewithryan) [inline src/classes/PREN_ProgramPlan_TDTM.cls]: did you intend to have the space here? I like the way you formatted line 77. Can we do the same here?

**Comment by xinnanlisf** [comment]
@ryanguest should be good now. Thanks!

**Comment by jbrandolinisf** [comment]
Hi @lianeyee, 

We do not need to change the description at all.  We changed an internal function of when the trigger fires (after instead of before), but nothing about the overall functionality.  Since the context was not mentioned, and the functionality did not change there is no update required for this.  

I believe there will be updates based on @xinnanlisf change, which is on the Program Enrollment object, if I remember correctly.

Hope this helps@

John 

PS - Referencing @lifewithryan and @ceiroa for awareness.

**Comment by dashattuck** [comment]
I tested this - and it looks good.  I was able to remove the program enrollment, and create a new program enrollment and see it auto-populate the program plan field correctly

**Comment by dashattuck** [comment]
successfully merged PR

**Comment by dashattuck** [comment]
Closed

**Comment by ceiroa** [comment]
Hey @jstvz, I don't see the text from this PR included in the latest beta release: https://github.com/SalesforceFoundation/HEDAP/releases. Could there maybe be an issue with the release notes functionality?

**Comment by jstvz** [comment]
@ceiroa I'm seeing it now (and also on the [build](https://mrbelvedereci.herokuapp.com/builds/43205/flows)). Let me know if that happens again.

**Pull request merged**